New Platform: Msp432

lebakassemmerl opened the pull request:
### New Platform Support for the MSP-EXP432P401R Eval-Board

This pull-request adds support for MSP432P401R chip from Texas Instruments.
The implementation is currently very, very basic, only the GPIO-module is properly implemented in order to create a panic-handler which blinks an LED and to be able to test a simple user-space application. 

If this PR will be accepted, the next step would be to implement the UART and timer modules for supporting debug-output and delays in userspace.

## Adding a Platform to Tock Repository

- [x] The hardware must be widely available. Generally that means the hardware platform can be purchased online.
- The port of Tock to the platform must include at least:
   - [x] Console support so that debug!() and printf() work.
   - [x] Timer support.
   - [x] GPIO support with interrupt functionality.
- [x] The contributor must be willing to maintain the platform, at least initially, and help test the platform for future releases.

alistair23 commented:
Thanks for the PR!!!

Would you mind rebasing on master? It looks like GitHub is confused and has included other patches in the PR. It makes it hard to review your changes.

Also the Tock porting guide gives some hints for porting: https://github.com/tock/tock/blob/master/doc/Porting.md#adding-a-platform-to-tock-repository
It also specifies that you need a working console (and timer but that might be less important) before the board is merged.

alistair23 commented:
It also looks like the tests are failing. Make sure to run at least `make ci-nosetup` locally. For new boards the CI check can be helpful in finding bugs (it checks all of your register offsets).

lebakassemmerl commented:
Hi!
Yes, of course. I will rebase it on master. Also I will make sure that the test won't fail.

Concerning the console and timer: I read this guide and somewhere at the beginning it said that after GPIO is working it's a good time to open a pull request. So should I close/delete this PR, fix my tests, get the console working and open a new one?

alistair23 commented:
Great! Thanks.

I think opening a PR makes sense. It lets everyone know what you are working on (so we don't duplicate efforts). It is also a great way to get feedback.

It might make sense to convert it to a draft PR until you have the tests and UART working though.

bradjc commented:
Very cool!

Can I ask what attracted you to this board/MCU?

lebakassemmerl commented:
Yeah, sure!

I'm doing my master thesis in porting Tock to a new platform, and my only 'requirement' was to choose a platform which is of course not supported yet and where a cheap evaluation board exists. In our university (FH Joanneum) we work a lot with all forms of TI products, so I had a look at their microcontroller boards. I chose this particular one because openOCD had good support for it, it contains a quite new Cortex-M4-based chip and the board costs only 20$.

lebakassemmerl commented:
Hi!

So far I managed it to get the UART in TX direction working in order to print panic- and debug-messages. Is it necessary for a first pull into the main-repo to get also the RX-functionality working? 

Also the GPIO implementation doesn't support interrupts at the moment. Is this also necessary to get merged?

alistair23 commented:
You don't need to have RX working, so that's fine. For GPIO I think that's also ok.

alistair23 commented on .vscode/launch.json:
```diff
@@ -15,6 +15,21 @@
         }
     ]
-}
+}
```
Can you keep the newline here

lebakassemmerl replied:
Yeah sure!

alistair23 commented on boards/msp_exp432p401r/Makefile:
```diff
@@ -0,0 +1,27 @@
+
+# APP=../../../libtock-rs/target/thumbv7em-none-eabi/tab/msp432/blink/cortex-m4.tbf
+# KERNEL=$(TOCK_ROOT_DIRECTORY)/target/$(TARGET)/debug/$(PLATFORM).elf
+# KERNEL_WITH_APP=$(TOCK_ROOT_DIRECTORY)/target/$(TARGET)/debug/$(PLATFORM)-app.elf
```
Can you remove the commented out code

alistair23 commented on boards/msp_exp432p401r/Makefile:
```diff
@@ -0,0 +1,27 @@
+program: $(TOCK_ROOT_DIRECTORY)target/$(TARGET)/debug/$(PLATFORM).elf
+	$(error See README.md and update this section accordingly)
+	# arm-none-eabi-objcopy --update-section .apps=$(APP) $(KERNEL) $(KERNEL_WITH_APP)
+	# $(OPENOCD) $(OPENOCD_OPTIONS) -c "init; reset halt; flash write_image erase $(KERNEL_WITH_APP); verify_image $(KERNEL_WITH_APP); reset; shutdown"
```
Same comment here about commented out code

alistair23 commented on boards/msp_exp432p401r/README.md:
```diff
@@ -0,0 +1,62 @@
+MSP432 Evaluation kit MSP432P401R MCU
+========================================================
```
Can you match the length of `===` with the header

alistair23 commented on boards/msp_exp432p401r/src/io.rs:
```diff
@@ -0,0 +1,58 @@
+pub static mut WRITER: Writer = Writer { initialized: false };
+
+impl Writer {
+    /// Indicate that USART has already been initialized.
```
UART?

alistair23 commented on boards/msp_exp432p401r/src/io.rs:
```diff
@@ -0,0 +1,58 @@
+    let led = &mut led::LedHigh::new(&mut msp432::gpio::PINS[LED1_PIN as usize]);
+    let writer = &mut WRITER;
+
+    //debug::panic_blink_forever(&mut [led]);
```
Remove commented out code.

alistair23 commented on boards/msp_exp432p401r/src/main.rs:
```diff
@@ -0,0 +1,163 @@
+#![no_std]
+#![cfg_attr(not(doc), no_main)]
+#![feature(asm, core_intrinsics)]
+// #![deny(missing_docs)]
```
This should be:

```
#![no_std]
// Disable this attribute when documenting, as a workaround for
// https://github.com/rust-lang/rust/issues/62184.
#![cfg_attr(not(doc), no_main)]
#![deny(missing_docs)]
```

Instead

alistair23 commented on boards/msp_exp432p401r/src/main.rs:
```diff
@@ -0,0 +1,163 @@
+// NOTE: This will cause the kernel to overwrite any existing apps when flashed!
+#[used]
+#[link_section = ".app.hack"]
+static APP_HACK: u8 = 0;
```
Do you mind explaining why you need this? Can apps not be flashed to the board?

lebakassemmerl replied:
I implemented it like the stm32f3discovery board, where the applications are flashed combined with the kernel. For me this seemed to be the easiest method to get the platform working at all. Is there a better/easier method for doing this?

alistair23 replied:
Most other boards allow you to flash the kernel and apps separately.You can just basically run the same flash command but with different addresses.

Something like this: https://github.com/tock/tock/blob/76b00caa8a7bb54778dcbbcdc9a1ce38148ef1fc/boards/nano33ble/README.md#programming-applications

lebakassemmerl replied:
Ah, I see!
Should I change this that I can flash the apps separately with openOCD (I guess this should be possible)? Or should I keep as it is now?

alistair23 replied:
Unless there is a reason you can't (the HW won't let you) you should allow flashing them separately. You can add a `make flash-app` target in the `Makefile` to make it easier to do if you want.

lebakassemmerl replied:
Ok, I added a new target in the makefile. Now it looks like the following:
```make
# Makefile for building the tock kernel for the MSP-EXP432P401 launchpad
#
TARGET=thumbv7em-none-eabi
PLATFORM=msp-exp432p401r

include ../Makefile.common

OPENOCD=openocd
OPENOCD_OPTIONS=-f openocd.cfg
APP_START_ADDR=0x20000
APP?=

.PHONY: flash-debug
flash-debug: $(TOCK_ROOT_DIRECTORY)target/$(TARGET)/debug/$(PLATFORM).elf
	$(OPENOCD) $(OPENOCD_OPTIONS) -c "init; reset halt; flash write_image erase $<; verify_image $<; reset; shutdown"

.PHONY: flash
flash: $(TOCK_ROOT_DIRECTORY)target/$(TARGET)/release/$(PLATFORM).elf
	$(OPENOCD) $(OPENOCD_OPTIONS) -c "init; reset halt; flash write_image erase $<; verify_image $<; reset; shutdown"

.PHONY: flash-app
flash-app:
	$(OPENOCD) $(OPENOCD_OPTIONS) -c "init; reset halt; flash write_image erase $(APP) $(APP_START_ADDR) bin; verify_image_checksum $(APP) $(APP_START_ADDR) bin; reset; shutdown"

.PHONY: program
program: $(TOCK_ROOT_DIRECTORY)target/$(TARGET)/debug/$(PLATFORM).elf
	$(error See README.md and update this section accordingly)`

Can I keep it like that? Or should I do something in a different way?

alistair23 commented on boards/msp_exp432p401r/src/main.rs:
```diff
@@ -0,0 +1,163 @@
+    msp432::gpio::PINS[msp432::gpio::PinNr::PJ_3 as usize].enable_primary_function();
+    msp432::cs::CS.set_clk_48mhz();
+    msp432::cs::CS.set_smclk_12mhz();
+
```
Here would be a good place to setup the debug GPIO with `kernel::debug::assign_gpios`

lebakassemmerl replied:
Sure, of course! I totally forgot about the GPIOs besides the LED.

alistair23 commented on chips/msp432/src/chip.rs:
```diff
@@ -0,0 +1,83 @@
+
+    fn sleep(&self) {
+        unsafe {
+            cortexm4::scb::unset_sleepdeep();
```
Why do you need this?

lebakassemmerl replied:
Ohh, another copy-paste bug from a different platform^^
I'll remove this line.

alistair23 commented on chips/msp432/src/cs.rs:
```diff
@@ -0,0 +1,269 @@
+const KEY: u32 = 0x695A;
+
+#[repr(C)]
+struct CsRegisters {
```
There is a `register_structs!` macro you can use for this

alistair23 commented on chips/msp432/src/cs.rs:
```diff
@@ -0,0 +1,269 @@
+
+#[allow(dead_code)]
+#[repr(u32)]
+enum DcoFrequency {
```
Is this used?

alistair23 commented on chips/msp432/src/cs.rs:
```diff
@@ -0,0 +1,269 @@
+        ClockSystem { registers: CS_BASE }
+    }
+
+    #[inline]
```
You probably can just let the compiler inline this, I don't think you need it to be explicit.

alistair23 commented on chips/msp432/src/lib.rs:
```diff
@@ -0,0 +1,140 @@
+#![crate_name = "msp432"]
+#![crate_type = "rlib"]
+#![feature(asm, const_fn, in_band_lifetimes)]
```
`in_band_lifetimes` have been removed from Tock, can you avoid re-introducing them?

lebakassemmerl replied:
Sure, I'll remove it!

alistair23 commented on chips/msp432/src/pcm.rs:
```diff
@@ -0,0 +1,113 @@
+const PCMKEY: u32 = 0x695A; // for unlocking PCMCTL0 and PCMCTL1
+
+#[repr(C)]
+struct PcmRegisters {
```
Same comment about `register_structs!`

alistair23 commented on chips/msp432/src/pcm.rs:
```diff
@@ -0,0 +1,113 @@
+
+#[allow(dead_code)]
+#[repr(u32)]
+enum ActiveMode {
```
Is this used?

alistair23 commented on chips/msp432/src/wdt.rs:
```diff
@@ -0,0 +1,69 @@
+const PASSWORD: u16 = 0x5A;
+
+#[repr(C)]
+struct WatchdogRegisters {
```
Same comment about `register_structs!`

alistair23 commented on chips/msp432/src/wdt.rs:
```diff
@@ -0,0 +1,69 @@
+        // TODO: implement the period for the watchdog
+        self.registers
+            .ctl
+            .modify(WDTCTL::WDTPW.val(PASSWORD) + WDTCTL::WDTHOLD::CLEAR);
```
Maybe add a call to `unimplemented!()` as this won't work

lebakassemmerl replied:
Ok!

alistair23 commented on chips/msp432/src/sysctl.rs:
```diff
@@ -0,0 +1,201 @@
+#[allow(dead_code)]
+const UNLKEY: u16 = 0x695A; // for unlocking IP protected secure zone
+#[allow(dead_code)]
+const UNLOCKED: u16 = 0xA596; // value if zone is unlocked
```
Why include these if they aren't used?

lebakassemmerl replied:
Of course I can remove them if you want, but these values are currently not used but are necessary for some register-accesses. I thought i define them although they are not used because then it's not necessary to look them up again in the datasheet.

alistair23 replied:
If you think they should be left in can you add a comment?

Someone will probably have to look at the datahseet to use them anyway so it still seems worth removing them if they are unused.

lebakassemmerl replied:
Ok, then I'll remove them (:

lebakassemmerl commented:
> Looks REALLY good! :)
> 
> I have added some comments, they should all be pretty small. Overall this looks really good. I didn't compare it to the datasheet but everything looks reasonable.
> 
> Do you mind addressesing the comments and then I'll have another look?

I think I have addressed all your comments with the latest commits. 
So, when you find the time, please have another look (:

ppannuto commented on boards/msp_exp432p401r/src/main.rs:
```diff
@@ -0,0 +1,165 @@
+
+/// RAM to be shared by all application processes.
+#[link_section = ".app_memory"]
+static mut APP_MEMORY: [u8; 32768] = [0; 32768];
```
Please update the `APP_MEMORY` to the new model in #1962 -- sorry these came somewhat concurrently :/

lebakassemmerl replied:
Sure, I will convert it to the new model

ppannuto commented on chips/msp432/src/cs.rs:
```diff
@@ -0,0 +1,275 @@
+        // select REFO nominal frequency: 0 = 32.768kHz, 1=128kHz
+        REFOFSEL OFFSET(15) NUMBITS(1)
+    ],
+    // status of the different clock-sources, if they are active or not
```
Please promote these to doc comments (`//` -> `///`) and add descriptors for the other register types as well so that documentation is generated for them.

lebakassemmerl replied:
Of course, I oversaw these comments.

ppannuto commented on chips/msp432/src/cs.rs:
```diff
@@ -0,0 +1,275 @@
+        ClockSystem { registers: CS_BASE }
+    }
+
+    fn unlock_registers(&self) {
+        self.registers.key.modify(CSKEY::KEY.val(KEY));
+    }
+
+    fn lock_registers(&self) {
+        // Every value except KEY written to the key register will perform the lock
+        self.registers.key.modify(CSKEY::KEY.val(0));
+    }
```
This is okay, but it would be better to use the [`PeripheralManager`](https://docs.tockos.org/kernel/common/peripherals/index.html) `before/after_peripheral_access` hooks, which allow an RAII-style interface that will automatically lock and unlock correctly.

lebakassemmerl replied:
Ok, I will try to convert it to this model!

lebakassemmerl replied:
I have now implemented the `PeripheralManagement` trait for the clock system, I hope I did it correct!

ppannuto commented on chips/msp432/src/flctl.rs:
```diff
@@ -0,0 +1,849 @@
+
+#[repr(u32)]
+#[derive(Copy, Clone)]
+pub enum WaitStates {
```
`///` comment here (what are the `WaitStates` used for?)

lebakassemmerl replied:
Ok!

ppannuto commented on chips/msp432/src/pcm.rs:
```diff
@@ -0,0 +1,188 @@
+    // currently not sure about the interface, so just implement a simple
+    // method for activating AM_LDO_VCORE1 to provide enough power for 48MHz
+    pub fn set_high_power(&self) {
+        while self.registers.ctl1.is_set(PCMCTL1::PMR_BUSY) {}
```
:/, another one for the #1552 wishlist

(nothing for you to do here; just want to flag it on that issue)

ppannuto commented on chips/msp432/src/lib.rs:
```diff
@@ -0,0 +1,140 @@
+#![crate_type = "rlib"]
+#![feature(asm, const_fn)]
+#![no_std]
+#![allow(unused_doc_comments)]
```
Remove this `allow`, it shouldn't be needed (though it was apparently used elsewhere; #1992...)

lebakassemmerl replied:
Ok!

ppannuto commented on chips/msp432/src/uart.rs:
```diff
@@ -0,0 +1,278 @@
+}
+
+#[rustfmt::skip]
+// Table out of the datahseet to correct the baudrate
```
```suggestion
// Table out of the datasheet to correct the baudrate
```

ppannuto commented on chips/msp432/src/uart.rs:
```diff
@@ -0,0 +1,278 @@
+        self.transmit();
+    }
+
+    fn transmit(&self) {
+        let idx = self.tx_index.get();
+
+        self.tx_buffer.map(|tx_buf| self.write_byte(tx_buf[idx]));
+        self.tx_index.set(idx + 1);
+    }
```
It looks like this is (asynchronously, good) transmitting one byte at a time. That's going to create a ton of interrupts. Is there no easy way to pass a length for the buffer to the uart peripheral?

lebakassemmerl replied:
This UART-module unfortunately doesn't have any kind of FIFO, it can only transmit one byte at a time. It could use a DMA-channel for avoiding these interrupts, but I thought for the first implementation I just go the easier way without DMA.

ppannuto replied:
I think that's reasonable as a start for basic functionality.

DMA should be on the shortlist of next implementation priorities once this is merged, however.

lebakassemmerl commented on boards/msp_exp432p401r/Makefile:
```diff
@@ -0,0 +1,23 @@
+OPENOCD=openocd
+OPENOCD_OPTIONS=-f openocd.cfg
+APP_START_ADDR=0x20000
+APP?=
```
It is just for flashing the tbf-file to the correct flash-location via openOCD, and with the `APP` variable you set the path of the tbf-file.

I added already a description and an example in the README. Should I add one in the makefile too?

ppannuto replied:
The README is probably fine for now.

alistair23 commented:
When you address a comment can you mark it as resolved?

lebakassemmerl commented:
> When you address a comment can you mark it as resolved?

Sure! I didn't know that I was supposed to mark the comments as resolved.

ppannuto commented on chips/msp432/src/cs.rs:
```diff
@@ -0,0 +1,308 @@
+    // Not sure about the interface, so for testing provide a function to set
+    // the master-clock to 48Mhz
+    pub fn set_mclk_48mhz(&self) {
+        self.before_peripheral_access(self.get_clock(), self.get_registers());
```
You should not need the explicit calls like this. The idea of the PeripheralManager is to use RAII, that is you create a `RegisterManager` object to get access to the registers. The manager object will automatically call `before` during creation and `after` when it goes out of scope.

ppannuto commented on chips/msp432/src/cs.rs:
```diff
@@ -0,0 +1,308 @@
+];
+
+pub struct ClockSystem {
+    registers: StaticRef<CsRegisters>,
```
```suggestion
```

With the PeripheralManager interface, this structure should not hold a reference to the MMIO registers. Register access needs to go through the peripheral manager instead.

lebakassemmerl replied:
Ahh ok, now I get it, thank you!
I think I use the `PeripheralMangager` now in the correct way

ppannuto commented:
I think you can move this from a Draft PR to a full-fledged one at the point :)

alistair23 commented on chips/msp432/src/uart.rs:
```diff
@@ -0,0 +1,278 @@
+        regs.ctlw0.modify(usci::UCAxCTLW0::UCSWRST::CLEAR);
+
+        // Enable receive interrupt
+        // self.registers.ie.modify(usci::UCAxIE::UCRXIE::SET);
```
Why isn't this enabled?

lebakassemmerl replied:
Because the receive part of the UART isn't implemented yet. I didn't want to enable an interrupt when the whole handling logic is not implemented. My thought was that this would cause only problems.

alistair23 replied:
Ah ok. I think it's best just to remove this then. Otherwise looks good :)

lebakassemmerl replied:
Ok, sure!

lebakassemmerl commented:
I implemented no a very first version for handling GPIO interrupts. However, there is a problem with this controller, it does NOT support detecting interrupts at the falling edge and at the rising edge simultaneously. Just for testing the implementation with the button-capsule I detect only falling edges where actually both should be detected. 

How should I solve this problem? Just 'ignore' the fact that both should be detected? Maybe `panic!()`? Or maybe any kind of software implementation (However I have no idea how this could be done within Tock)?

hudson-ayers commented:
> I implemented no a very first version for handling GPIO interrupts. However, there is a problem with this controller, it does NOT support detecting interrupts at the falling edge and at the rising edge simultaneously. Just for testing the implementation with the button-capsule I detect only falling edges where actually both should be detected.
> 
> How should I solve this problem? Just 'ignore' the fact that both should be detected? Maybe `panic!()`? Or maybe any kind of software implementation (However I have no idea how this could be done within Tock)?

Regarding a software implementation, I think it would basically involve re-configuring a pin to check for the other edge type each time an interrupt arrived for that pin. This would just be done in the chip GPIO driver -- basically when a falling edge interrupt arrives, check that the pin is still low, and if so configure that pin to look for a rising edge interrupt, and vice-versa. Unfortunately it would have to take place in the bottom half interrupt handler, which means there is the possibility for large enough delays that interrupts could be missed, but I think this is still a better solution than only handling interrupts in one direction.

lebakassemmerl commented:
> > I implemented no a very first version for handling GPIO interrupts. However, there is a problem with this controller, it does NOT support detecting interrupts at the falling edge and at the rising edge simultaneously. Just for testing the implementation with the button-capsule I detect only falling edges where actually both should be detected.
> > How should I solve this problem? Just 'ignore' the fact that both should be detected? Maybe `panic!()`? Or maybe any kind of software implementation (However I have no idea how this could be done within Tock)?
> 
> Regarding a software implementation, I think it would basically involve re-configuring a pin to check for the other edge type each time an interrupt arrived for that pin. This would just be done in the chip GPIO driver -- basically when a falling edge interrupt arrives, check that the pin is still low, and if so configure that pin to look for a rising edge interrupt, and vice-versa. Unfortunately it would have to take place in the bottom half interrupt handler, which means there is the possibility for large enough delays that interrupts could be missed, but I think this is still a better solution than only handling interrupts in one direction.

Ok, thanks!
Then I will try to implement a first software based approach for detecting both edges simultaneously.

lebakassemmerl commented:
I have now implemented a first approach for detecting both interrupts by software. Can somebody please have a look on it if the implementation makes sense?
I tested it within the debugger, the basic functionality seems to work, after a rising edge it detects a falling one and vice versa. However I am not sure if I didn't make any subtle mistakes in this implementation.

hudson-ayers commented:
> I have now implemented a first approach for detecting both interrupts by software. Can somebody please have a look on it if the implementation makes sense?
> I tested it within the debugger, the basic functionality seems to work, after a rising edge it detects a falling one and vice versa. However I am not sure if I didn't make any subtle mistakes in this implementation.

This seems like a reasonable approach! Have you tried the button example?

lebakassemmerl commented:
> > I have now implemented a first approach for detecting both interrupts by software. Can somebody please have a look on it if the implementation makes sense?
> > I tested it within the debugger, the basic functionality seems to work, after a rising edge it detects a falling one and vice versa. However I am not sure if I didn't make any subtle mistakes in this implementation.
> 
> This seems like a reasonable approach! Have you tried the button example?

Yes I did and it works. However, it also worked when I configured the controller to detect only the falling or rising edge. I think that's due to the bouncing of the button itself, maybe the button example isn't the best one to test this implementation.

lebakassemmerl commented:
I have now implemented a first version of an alarm timer. Is it also necessary to implement the `Timer` trait for accepting the PR? Or is the `Alarm` trait sufficient?

bradjc commented:
As long as the blink app works, and that only needs alarm.

lebakassemmerl commented:
> As long as the blink app works, and that only needs alarm.

Ok, thanks!
With the latest commits the blink-app works now.

lebakassemmerl commented:
Thanks for approving my changes!
Am I supposed to do something else in order to get the PR accepted? Or is the current implementation sufficient for now?

alistair23 commented:
I think you have everything you need. You just need to wait for some more reviews

bradjc commented on chips/msp432/src/chip.rs:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    fn watchdog(&self) -> &Self::WatchDog {
+        unsafe { &wdt::WDT }
```
This is concerning, why is it unsafe to use the watchdog here?

lebakassemmerl replied:
Because the rust-compiler won't let me return a reference to a static mutable variable. Because there is only 1 watchdog instances I thought this should not be a big problem although it is marked as unsafe.

hudson-ayers replied:
I think that rather than initializing the watchdog timer as a global, it should just be a field of the MSP432 struct, and be initialized in the constructor for that struct.

Using it this way opens the door for future code in the kernel to take a simultaneous mutable reference to the WDT, violating Rust's aliasing guarantees.

lebakassemmerl replied:
Ahh, I see!
Then I create a field within the Msp432 struct for the watchdog and delete the global instance.

hudson-ayers replied:
Is it necessary to stop it immediately at startup? Could that just be done as part of the call to `msp432::new()`? Either way, I think that temporarily allocating a short lived instance on the stack (as you are doing) is also fine.

lebakassemmerl replied:
I think it is necessary to stop it immediately at startup, because the interval for it is only ~10ms and depending on what is done before the Msp432 struct is instantiated it could be that the watchdog resets the system.

lebakassemmerl commented:
With the latest commit I added some missing lifetime specifiers in the GPIO module. However, since the struct `PinJ` basically doesn't need any lifetime (because it doesn't support interrupts), the macro I created doesn't work any more. For this reason I added a `PhantomData<'a ()>` member to  `PinJ` which makes it behave as it would have a lifetime. Is this the intended way to go or is there a better one?

bradjc commented:
bors r+
